snuba_sdk/query_validation.py

```diff
@@ -0,0 +1,100 @@
+from __future__ import annotations
+
+from typing import Set
+
+# Import the modules due to sphinx autodoc problems
+# https://github.com/agronholm/sphinx-autodoc-typehints#dealing-with-circular-imports
+from snuba_sdk import query as main
+from snuba_sdk import query_visitors as qvisitors
+from snuba_sdk.column import Column
+from snuba_sdk.entity import Entity
+from snuba_sdk.expressions import Expression
+from snuba_sdk.function import CurriedFunction
+from snuba_sdk.relationships import Join
+
+
+class InvalidMatchError(Exception):
+    pass
+
+
+def validate_match(
+    query: main.Query, column_finder: qvisitors.ExpressionSearcher
+) -> None:
+    """
+    Perform validation related to the match clause of the query. Currently that takes two forms:
+    validate that the columns referenced in the query are valid columns based on the match clause.
+
+    :param query: The query to be validated.
+    :param column_finder: An ExpressionSearcher used to find all the columns in the query.
+    """
+    all_columns = column_finder.visit(query)
+    if isinstance(query.match, main.Query):
+        _validate_subquery(query.match, all_columns)
+        validate_match(query.match, column_finder)
```

Comment:
What does it mean if query.match is itself a query (neither an entity or a join)? Why do we call validate_match recursively in this scenario?

Reply:
It means that the MATCH is a subquery. I had this here because the inner query needs to be validated, but I just realized the inner query is validated when it's added so we don't need the recursion here.